### PR TITLE
Add board support for STM32F401 Discovery

### PR DESCRIPTION
**Pull request by cajt**
Added a BSP for the F401 Discovery Board.

Notes:
- Completed basic testing with the included examples, including the USART, Gyro and Accelerometer.
- Board is obsolete according to ST.

### CODE REVIEW

**Comment by rleh** [comment]
The CI complains about two things:
- Please add the new example folder to the appropriate CI job in `.github/workflows/linux.yml`.
- Please run `python3 tools/scripts/synchronize_docs.py` to synchronize the documentation.

File: examples/stm32f401_discovery/uart/main.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include <modm/board.hpp>
+
+// ----------------------------------------------------------------------------
+/**
+ * Very basic example of USART usage.
+ * The ASCII sequence 'A', 'B', 'C, ... , 'Z', 'A', 'B', 'C', ...
```
Comment (rleh) [inline examples/stm32f401_discovery/uart/main.cpp]:
```suggestion
 * The ASCII sequence 'A', 'B', 'C', ... , 'Z', 'A', 'B', 'C', ...
```

**Review by rleh — APPROVED** [review]
We really need to take care of deduplicating the examples somehow, but that shouldn't stop this pull request.

File: examples/stm32f401_discovery/uart/main.cpp (hunk 1 above)
Comment (cajt) [reply examples/stm32f401_discovery/uart/main.cpp]: done.

**Comment by cajt** [comment]
> We really need to take care of deduplicating the examples somehow, but that shouldn't stop this pull request.

From the abstraction side, the boards looked almost identical and could run mostly unmodified code (only difference was number of LEDs and their names). 

The way I used the examples was basically a unit test, to check is the BSP is working. For use as code examples the level of abstraction is fine imho, and kinda what I would expect. To ease dealing with testing, a more abstract and generalized way of writing these might help. Especially for more elaborate tests. But that things can't do double duty as example and unit-test anymore (if that was ever intended).

**Review by salkinium — APPROVED** [review]
Thanks!

(I'll squash your commits and merge it later)

**Pull request merged**